crates/bevy_ecs/src/schedule_v3/graph/utils.rs

```diff
@@ -0,0 +1,197 @@
+}
+
+/// Converts a 1D array index into a 2D row-major pair of indices.
+pub(crate) fn row_col(index: usize, num_cols: usize) -> (usize, usize) {
+    (index / num_cols, index % num_cols)
+}
+
+pub(crate) struct CheckGraphResults<V> {
+    // Pairs of nodes that have a path connecting them.
+    pub(crate) connected: HashSet<(V, V)>,
+    // Pairs of nodes that don't have a path connecting them.
+    pub(crate) disconnected: HashSet<(V, V)>,
+    // Edges that are redundant because a longer path exists.
+    pub(crate) transitive_edges: Vec<(V, V)>,
+    // Boolean reachability matrix for the graph.
+    pub(crate) reachable: FixedBitSet,
+    // Variant of the graph with the fewest possible edges.
+    #[allow(dead_code)]
+    pub(crate) tred: DiGraphMap<V, ()>,
+    // Variant of the graph with the most possible edges.
+    #[allow(dead_code)]
+    pub(crate) tcls: DiGraphMap<V, ()>,
+}
+
+impl<V: NodeTrait + Debug> Default for CheckGraphResults<V> {
+    fn default() -> Self {
+        Self {
+            connected: HashSet::new(),
+            disconnected: HashSet::new(),
+            transitive_edges: Vec::new(),
+            reachable: FixedBitSet::new(),
+            tred: DiGraphMap::new(),
+            tcls: DiGraphMap::new(),
+        }
+    }
+}
+
+pub(crate) fn check_graph<V>(
+    graph: &DiGraphMap<V, ()>,
+    topological_order: &[V],
+) -> CheckGraphResults<V>
+where
+    V: NodeTrait + Debug,
+{
+    if graph.node_count() == 0 {
+        return CheckGraphResults::default();
+    }
+
+    let n = graph.node_count();
+    let mut map = HashMap::with_capacity(n);
+    let mut tsorted = DiGraphMap::<V, ()>::new();
+    // iterate nodes in topological order
+    for (i, &node) in topological_order.iter().enumerate() {
+        map.insert(node, i);
+        tsorted.add_node(node.clone());
+        // insert nodes as successors to their predecessors
+        for pred in graph.neighbors_directed(node, Direction::Incoming) {
+            tsorted.add_edge(pred, node, ());
+        }
+    }
+
+    let mut tred = DiGraphMap::<V, ()>::new();
+    let mut tcls = DiGraphMap::<V, ()>::new();
+
+    let mut connected = HashSet::new();
+    let mut disconnected = HashSet::new();
+    let mut transitive_edges = Vec::new();
+
+    let mut visited = FixedBitSet::with_capacity(n);
+    let mut reachable = FixedBitSet::with_capacity(n * n);
+
+    // iterate nodes in topological order
+    for node in tsorted.nodes() {
+        tred.add_node(node);
+        tcls.add_node(node);
+    }
+
+    // iterate nodes in reverse topological order
+    for a in tsorted.nodes().rev() {
+        let index_a = *map.get(&a).unwrap();
+        // iterate their successors in topological order
+        for b in tsorted.neighbors_directed(a, Direction::Outgoing) {
+            let index_b = *map.get(&b).unwrap();
+            debug_assert!(index_a < index_b);
+            if !visited[index_b] {
+                // edge <a, b> is not redundant
```

Comment:
I don't understand this comment.

Reply:
This function computes the transitive reduction and closure of a graph (plus extras). This edge cannot be removed.